index.bs

```diff
@@ -8972,6 +9018,371 @@ worker=] algorithm:
+
+  1. Let |partition key| be |partition spec|.
+
+1. For each |name| → |default value| in the [=default values for storage partition key attributes=]:
+
+  1. If |partition key|[|name|] does not [=map/exist=]:
+
+    1. [=map/Set=] |partition key|[|name|] to |default value|.
+
+1. If |partition key|["<code>sourceOrigin</code>"] does not [=map/exist=]:
+
+  1. If |cookie domain| is null, then:
+
+    1. Return [=error=] with [=error code=] [=unknown source origin=].
+
+  1. [=map/Set=] |partition key|["<code>sourceOrigin</code>"] to |cookie domain|.
+
+1. Return [=success=] with data |partition key|.
+
+</div>
+
+<div algorithm>
+To <dfn>get the cookie store</dfn> given |storage partition key|:
+
+1. If a [=cookie store=] |match| exists for |storage partition key|:
+
+  1. Return [=success=] with data |match|.
+
+1. Return [=error=] with [=error code=] [=no such storage partition=].
+
+</div>
+
+<div algorithm>
+To <dfn>match cookie</dfn> given |stored cookie| and |filter|:
+
+1. For each |name| → |value| in |filter|:
+
+  1. If |name| is "<code>value</code>":
+
+    1. Set |value| to [=deserialize protocol bytes=] with |value|.
+
+  1. Let |field name| be the field name corresponding to the JSON key |name| in
+    the [=table for cookie conversion=].
+
+  1. If |stored cookie|[|field name|] does not equal |value|:
+
+    1. Return false.
+
+1. Return true.
+
+</div>
+
+<div algorithm>
+To <dfn>get matching cookies</dfn> given |cookie store| and |filter|:
+
+1. Let |cookies| be a new list.
+
+1. For each |stored cookie| in |cookie store|:
+
+  1. If [=match cookie=] with |stored cookie| and |filter| is true:
+
+    1. Append the result of [=serialize cookie=] given |stored cookie| to |cookies|.
+
+1. Return |cookies|.
+
+</div>
+
+### Commands ### {#module-storage-commands}
+
+#### The storage.getCookies Command ####  {#command-storage-getCookies}
+
+The <dfn export for=commands>storage.getCookies</dfn> command retrieves zero or more [=cookies=] which [=match cookie|match=] a set of provided parameters.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+    <pre class="cddl remote-cddl">
+        storage.GetCookies = {
+          method: "storage.getCookies",
+          params: storage.GetCookiesParameters
+        }
+
+        <!--
+        Modifications to this definition should be reflected in
+        `network.Cookie`, `network.SetCookieHeader`, and `storage.PartialCookie`.
+        -->
+        storage.CookieFilter = {
+          ? name: text,
+          ? value: network.BytesValue,
+          ? domain: text,
+          ? path: text,
+          ? size: js-uint,
+          ? httpOnly: bool,
+          ? secure: bool,
+          ? sameSite: network.SameSite,
+          ? expiry: js-uint,
+          Extensible,
+        }
+
+        storage.GetCookiesParameters = {
+          cookie: storage.CookieFilter,
+          ? partition: storage.PartitionKey / browsingContext.BrowsingContext,
```

Comment:
So when we have this kind of choice we're preferring named types over structural typing. So in this design the options should be something like:
```
storage.GetCookiesParameters = {
[…]
? partition: storage.BrowsingContextPartition / storage.StorageKeyPartition
}

storage.BrowsingContextPartition = {
  type: "context",
  context: browsingContext.BrowsingContext
 }
 
 storage.StorageKeyPartition = {
   type: "storage-key",
   ? userContext: text,
   ? sourceOrigin: text,
 }
 ```

Reply:
That seems more amenable to future extensions, though I'm curious if there are any other motivations.

I think referring to these as types of "Partitions" could be misleading because it implies that they're mutually exclusive. As far as I understand, one could use either type to describe the same storage partition. This would be more clear to me if we qualified these types with a suffix like "Descriptor." The resultant verbosity could be reduced with a new group:

```diff
 storage.GetCookiesParameters = {
   […]
-  ? partition: storage.BrowsingContextPartition / storage.StorageKeyPartition
+  ? partition: storage.PartitionDescriptor
 }
+
+storage.PartitionDescriptor = (
+  storage.BrowsingContextPartitionDescriptor /
+  storage.StorageKeyPartitionDescriptor
+)
 
-storage.BrowsingContextPartition = {
+storage.BrowsingContextPartitionDescriptor = {
   type: "context",
   context: browsingContext.BrowsingContext
 }
 
-storage.StorageKeyPartition = {
+storage.StorageKeyPartitionDescriptor = {
   type: "storageKey",
   ? userContext: text,
   ? sourceOrigin: text,
+  Extensible,
 }
```

(Also, whether we call it `StorageKeyPartition` or `StorageKeyPartitionDescriptor`, I think [our previous discussion about allowing additional fields](https://github.com/w3c/webdriver-bidi/pull/501#discussion_r1378734361) still applies.)

This leaves room for a type name like `storage.PartitionKey` as the return value for "set cookie" and "delete cookie". I think there's value in a distinct type (rather than using `storage.StorageKeyPartition`/`storage.StorageKeyPartitionDescriptor`) because the `type` field is superfluous in that context and because the `userContext` and `sourceOrigin` field would not be optional.

```diff
 storage.PartitionKey {
-    ? userContext: text,
+    userContext: text,
-    ? sourceOrigin: text,
+    sourceOrigin: text,
     Extensible,
 }
```

That design would also involve a delineation between remote end definitions ( `storage.PartitionDescriptor`, `storage.BrowsingContextPartitionDescriptor`, and `storage.StorageKeyPartitionDescriptor`) and local end definitions (`storage.PartitionKey`).

What do you think?

Reply:
> That seems more amenable to future extensions, though I'm curious if there are any other motivations.

Mostly that. I don't know that everyone agrees with the design decision, but in my opinion it helps make parsing into program types unambiguous, and as you say makes future extensions easier.

> What do you think?

I don't have a strong feeling about internal type names, so your suggestions seem fine. I agree that separating out the type that we return from the input type makes sense given that they have different validity constraints.